### PR TITLE
Migrating the 'resources' metadata payload to component

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

Migrating the  `resources` metadata to a component.

### Describe how to test/QA your changes

Run the new version of the agent and check that the `resources` payload is still sent with the correct format.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/metadata/runner/runner.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+	for {
+		go func(intervalChan chan time.Duration) {
+			intervalChan <- p(ctx, r.serializer)
+		}(intervalChan)
+
+		select {
+		case interval = <-intervalChan:
+		case <-r.stopChan:
+			cancel()
```
Comment (ogaca-dd) [inline comp/metadata/runner/runner.go:117]: Note: This is strictly not needed as `cancel` is called [here](https://github.com/DataDog/datadog-agent/pull/18099/files#diff-effd0202d7ac04db2a5ed611246a121d1d512c137dfb2288c9e1b250427d7288R104)

**Review by ogaca-dd — CHANGES_REQUESTED** [review]
Great PR!
I have selected `Request changes` as there are some important questions to answer before merging this PR.

File: comp/core/config/config.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func (c *cfg) Warnings() *config.Warnings {
 	return c.warnings
 }
 
-func newMock(deps mockDependencies, t testing.TB) (Component, error) {
+func newMock(deps mockDependencies, t testing.TB) (Mock, error) {
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/core/config/config.go:69]: I forgot to update this after our latest discussion around mocking config. Thanks for catching it.

File: comp/metadata/resources/resources.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/DataDog/datadog-agent/pkg/gohai/processes"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/util/hostname"
+	"go.uber.org/fx"
+)
+
+var (
+	defaultCollectInterval = 300 * time.Second
+	providerName           = "resources"
+)
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/resources/resources.go:28]: Totally, copy paste from the old code :+1:

File: comp/metadata/resources/resources.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+type provides struct {
+	fx.Out
+
+	Comp     Component
+	Provider runner.Provider
+}
+
+func newResourcesProvider(deps dependencies) provides {
+	// By default, the 'resources' metadata is only enabled on Linux. Users can manually enable it on darwin
+	// platform. This is legacy behavior from Agent V5.
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/resources/resources.go:53]: It used to be defined here: https://github.com/DataDog/datadog-agent/pull/18099/files/7f6558f67796703bada78f545682bf85bf2e73de#diff-96fff4e95d733d53c0d6fad999b9b29d126e5ffc020f3cf96194f2ea161f3b8aL49

File: comp/metadata/resources/resources_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		{
+			Name:     "resources",
+			Interval: 21,
+		},
+	})
+
+	ret := newResourcesProvider(
+		dependencies{
+			Log:    fxutil.Test[log.Component](t, log.MockModule),
+			Config: config.Component(c),
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/resources/resources_test.go:61]: No we needed to cast to satisfy the compiler. But I remove the logic to edit the conf after init, so this cast is no longer needed.

File: comp/metadata/runner/runner.go [hunk 6]
```diff
@@ -12,49 +12,67 @@ import (
 
 	"github.com/DataDog/datadog-agent/comp/core/config"
 	"github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/util"
 	"go.uber.org/fx"
 )
 
-type MetadataProvider func(context.Context) time.Duration
+type MetadataProvider func(context.Context, serializer.MetricSerializer) time.Duration
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/runner/runner.go:20]: I forgot to edit this :+1:. Now that we have the serializer in FX we don't need the runner to centralize the serializer initalization.

File: cmd/agent/subcommands/run/command.go [hunk 7]
```diff
@@ -276,6 +284,25 @@ func getSharedFxOption() fx.Option {
...
+		fx.Provide(func(config config.Component, log log.Component, sharedForwarder defaultforwarder.Component) (*aggregator.AgentDemultiplexer, error) {
+			opts := aggregator.DefaultAgentDemultiplexerOptions()
+			opts.EnableNoAggregationPipeline = config.GetBool("dogstatsd_no_aggregation_pipeline")
+			opts.UseDogstatsdContextLimiter = true
+			opts.DogstatsdMaxMetricsTags = config.GetInt("dogstatsd_max_metrics_tags")
+			hostnameDetected, err := hostname.Get(context.TODO())
+			if err != nil {
+				return nil, log.Errorf("Error while getting hostname, exiting: %v", err)
+			}
+			return aggregator.InitAndStartAgentDemultiplexer(log, sharedForwarder, opts, hostnameDetected), nil
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply cmd/agent/subcommands/run/command.go:299]: Yes, this should be fine unless I miss something.

File: comp/metadata/resources/resources_non_supported.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	"go.uber.org/fx"
+)
+
+type provides struct {
+	fx.Out
+
+	Comp Component
+}
+
+func newResourcesProvider(log log.Component, config config.Component) provides {
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/resources/resources_non_supported.go:22]: I think we do need the comp in `provides` so FX holds a ref to the `resources` struct preventing the GC to collect it, no ?

[the remark this answers is not in the document]
Comment (ogaca-dd) [reply comp/metadata/resources/resources_non_supported.go:22]: What do you mean by `resources` struct?

[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/resources/resources_non_supported.go:22]: Actually  we need this type or FX will error-out:

Ex:
```
Failed: cannot provide function "github.com/DataDog/datadog-agent/comp/metadata/resources".newResourcesProvider (C:/dev/go/src/github.com/DataDog/datadog-agent/comp/metadata/resources/resources_non_supported.go:21): func(log.Component, config.Component) resources.provides must provide at least one non-error type
```

**Review by ogaca-dd — APPROVED** [review]

**Review by truthbk — APPROVED** [review]

**Pull request merged**